### DISCUSSION TITLE
Airflow event listeners for for dagrun

### DISCUSSION THREAD

**Discussion by mishrasangeeta87** [apache/airflow#30701] [General] [answered]
Hi All,

I was going through the listener documentation : https://airflow.apache.org/docs/apache-airflow/2.5.3/administration-and-deployment/listeners.html and reused the sample code mentioned in the doc: https://airflow.apache.org/docs/apache-airflow/stable/_modules/airflow/example_dags/plugins/event_listener.html

During the Scheduler start up I got a PluginValidation error for on_dag_run events.

` raise PluginValidationError(
pluggy._manager.PluginValidationError: Plugin 'airflow.example_dags.plugins.event_listener' for hook 'on_dag_run_failed'
hookimpl definition: on_dag_run_failed(dag_run: 'DagRun', message: 'str')
Argument(s) {'message'} are declared in the hookimpl but can not be found in the hookspec`

I made changes to conform to the spec. Restarted the components. Plugin was now visible under http://localhost:8080/plugin

I triggered one DAG and was able to see task instance event messages from the listener in the log. However nothing related to dag run events were being captured in the logs. Do let me know if you have any experience implementing dag run event listeners.

Thanks

**Comment by potiuk** [comment]
Maybe @mobuchowski can help, I think he was dealing with the listener interface recently . (BTW. you need to also specify Airflow Version you use @mishrasangeeta87 -there were some changes to the listener interface recently)

**Reply by mishrasangeeta87** [reply]
airflow version being used is 2.5.3

**Reply by mobuchowski** [reply]
@mishrasangeeta87 the argument is specified as `msg` not `message`:

```
@hookspec
def on_dag_run_failed(dag_run: DagRun, msg: str):
    """Called when dag run state changes to FAIL."""
```

**Reply by mishrasangeeta87** [reply]
@mobuchowski  Once the spec was failing I modified it to msg and then the plugin was visible. But still dag_run messages didn't show up in the logs.

Pasting the plugin code  and logs for your reference which is not capturing dag run event logs:
**listener_plugin.py** https://gist.github.com/mishrasangeeta87/1e81095afd45df2399c5517494c92ba5
 **event_listener.py** https://gist.github.com/mishrasangeeta87/6404a3ca63fcd370a5719190273df310
**hello_world_dag.py** https://gist.github.com/mishrasangeeta87/2953e3900b7cf3599167bb3aea2b60dc
**Task logs** https://gist.github.com/mishrasangeeta87/a9b04adf6971b2b5171e336d3a3beea8

**Reply by mobuchowski** [reply]
@mishrasangeeta87 try using Airflow logging instead of prints: https://github.com/apache/airflow/blob/47b67f13da667de4eb69233404d4d5a33574dd58/airflow/utils/log/logging_mixin.py#L63

**Reply by mishrasangeeta87** [reply]
@mobuchowski : I tried to use LoggingMixin in my _event_listener.py_ code  https://gist.github.com/mishrasangeeta87/c465d1898141a653382eae7a99dec633. For this I had to wrap entire listener code under class and add self parameter in all the event specs. Due to the additional parameter already working TaskInstance event listener has also stopped working. Could you redirect me to some samples wherein we can use Airflow logging without using LoggingMixin as a super class? Wasn't able to find code using  LoggingMixin without wrapping the code in a class in the airflow codebase.

**Comment by mobuchowski** [comment, marked as answer]
@mishrasangeeta87 

>For this I had to wrap entire listener code under class and add self parameter in all the event specs. Due to the additional parameter already working TaskInstance event listener has also stopped working.

This was already fixed, but fix will be only in 2.6 😞 

>Could you redirect me to some samples wherein we can use Airflow logging without using LoggingMixin as a super class? 

I think better option until 2.6 is that you should call the class from the listener as a function. That class would use LoggingMixin.

**Reply by asluis3** [reply]
Hey @mobuchowski would you mind giving me a small code snippet example of what you mean here? I'm having difficulties with the workaround for this bug. I'm also on 2.5.3—thank you!
